### ISSUE TITLE
Closing the database connection pool does not clear cursors using those connections, so dropping a database can result in a cursor that exists (but is unusable)

### ISSUE THREAD

**Issue by danielkauffman** [odoo/odoo#38561]
**Impacted versions:** Odoo 13

**Steps to reproduce:**

Drop the database. Note that request.cursor exists (but is unusable) after the database has been dropped.

**Current behavior:**

When dropping a database via the web interface, the database is dropped and the cursor is cleared, see https://github.com/odoo/odoo/blob/13.0/addons/web/controllers/main.py#L1033

            dispatch_rpc('db','drop', [master_pwd, name])
            request._cr = None  # dropping a database leads to an unusable cursor

However, when dropping a database via RPC, the database is dropped but the cursor is not cleared.

**Expected behavior:**

The cursor should be cleared when the database is dropped, regardless of the method for dropping the database.

The underlying issue is that odoo.sql_db.ConnectionPool().close_all() closes the underlying connections but does not clear cursors using those connections.

An easy fix for my use case would be to modify odoo.sql_db.close_db() to clear the cursor just before closing the connection to the database, but that would not resolve the underlying issue with the disconnect between the state of the ConnectionPool and the state of the request.cursor.

**Comment by pedrobaeza** [comment]
You are not indicating how you are dropping the DB via XML-RPC. Any way, I think just starting a new RPC client should be enough.

**Comment by danielkauffman** [comment]
In my case, I'm using JSON RPC to execute the drop method on the db object, which I think translates to odoo.http.dispatch_rpc('db', 'drop', params).

You are correct that catching the exception in the RPC client, and then reconnecting, allows the RPC client to continue without further exceptions.

The real issue is that the state of the ConnectionPool and the state of the request cursor are able to drift out of sync. The code for the cursor currently assumes that these will stay in sync, and throws an exception when they do not. **I think the cursor should gracefully handle cases where the database is dropped, or where the connection to the database is closed.**

The code for dropping a database via the web interface currently hacks around the issue by manually clearing the request cursor. If the solution is to hack around the issue, then I think this hack would be a better fit for odoo.sql_db.close_db().

But even better would be setting some sort of flag on the connection to the database, so that when the connection is closed, the cursor is able to detect the change in state for the underlying connection, and respond accordingly. That would allow the cursor to distinguish between the database connection being closed by Odoo, and the database connection being closed for some other reason (for example, the database restarting, or database-initiated garbage collection of old connections). But the exact behavior for the cursor under different conditions would need some discussion.

**Comment by pedrobaeza** [comment]
Well, I can only ping @odony for his opinion on this, as this is a very specific corner case as dropping databases is not the regular operation mode.

**Comment by danielkauffman** [comment]
Yes, dropping the database is definitely a corner case. ;-) The reason I reported this is because the corner case exposes an issue with the ConnectionPool state vs the request cursor state. And those being able to drift out of sync is likely to affect other cases as well.

**Comment by odony** [comment]
It would be easier to discuss the expected behavior based on some code (e.g. a PR, ideally with a test), to make it more concrete ;-)

**Comment by danielkauffman** [comment]
Quite right! I'll try to knock out a PR next week.

**Comment by C3POdoo** [comment]
Dear @danielkauffman,

Thank you for your report but we are closing it due to inactivity.
We apology if we could not look at your request in time.
If your report still makes sense, don't hesitate to reopen a new one. We will try to check it as soon as possible.

_This is an automated message._
